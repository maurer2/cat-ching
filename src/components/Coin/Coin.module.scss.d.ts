import globalClassNames from '../../style.d';
declare const classNames: typeof globalClassNames & {
  readonly 'container': 'container';
  readonly 'image-container': 'image-container';
  readonly 'title': 'title';
  readonly 'title--hidden': 'title--hidden';
  readonly 'image': 'image';
  readonly 'button-group': 'button-group';
  readonly 'button': 'button';
  readonly 'hidden-text': 'hidden-text';
};
export = classNames;
