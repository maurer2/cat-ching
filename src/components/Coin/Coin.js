import React, { useState } from 'react';
import uuidv1 from 'uuid/v1';
import PropTypes from 'prop-types';

import style from './Coin.module.scss';

const Button = ({ handleOnClick, id, children }) => (
  <button
    className={style.button}
    onClick={handleOnClick}
    id={id}
    type="button"
  >
    { children }
  </button>
);

function Coin({ name, image, value, handleAmountChange }) {
  const [hintIsVisible, setHintIsVisible] = useState(false);
  const htmlId = React.useRef(uuidv1());

  function addAmount() {
    handleAmountChange(value.valueInCents);
  }

  function removeAmount() {
    handleAmountChange(-value.valueInCents);
  }

  function toggleHintVisibility() {
    setHintIsVisible(!hintIsVisible);
  }

  return (
    <fieldset className={style.container}>
      <label className={style.header} htmlFor={htmlId.current}>
        {hintIsVisible && (
          <span className={style.title}>
            {name}
          </span>
        )}
        <img className={style.image} src={`/images/${image}`} alt="" />
      </label>
      <div className={style.buttonGroup}>
        <Button handleOnClick={addAmount} id={htmlId.current}>
          Add amount
        </Button>
        <Button handleOnClick={removeAmount}>
          Remove amount
        </Button>
        <Button handleOnClick={toggleHintVisibility}>
          {hintIsVisible ? 'Hide' : 'Show' }
          hint
        </Button>
      </div>
    </fieldset>
  );
}

export default Coin;

const { string, number, func, shape, node } = PropTypes;

Coin.propTypes = {
  name: string.isRequired,
  value: shape({
    value: number.isRequired,
    currency: string.isRequired,
    valueInCents: number.isRequired,
    valueFormatted: string.isRequired,
  }).isRequired,
  image: string.isRequired,
  handleAmountChange: func.isRequired,
};

Button.propTypes = {
  handleOnClick: func.isRequired,
  id: string,
  children: node.isRequired,
};

Button.defaultProps = {
  id: undefined,
};
