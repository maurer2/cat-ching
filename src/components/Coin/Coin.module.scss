.container {
  margin: 0;
  padding: 0;
  border: 0;
}

.header {
  display: flex;
  flex-direction: column;
  text-align: center;
}

.title {
  font-weight: bold;
}

.image {
  display: block;
  margin: 1rem auto;
  object-fit: contain;
}

.buttonGroup {
  display: flex;
  margin: 0 -0.5rem;
  justify-content: space-evenly;
}

.button {
  margin-left: 0.25rem;
  margin-right: 0.25rem;
  padding: 0.5rem;
  border: 1px solid black;
  border-radius: 0;
}
