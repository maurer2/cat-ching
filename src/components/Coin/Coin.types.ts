import type Money from '../../types/Money';
import type { Coin } from '../../types/Coin';

export type CoinProps = {
  coin: Coin,
  onAddAmount: (amount: Money) => void,
  onSubtractAmount: (amount: Money) => void,
};

export type ButtonProps = {
  handleOnClick: () => void
};
