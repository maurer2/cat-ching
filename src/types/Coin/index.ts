export type { CoinData, Coin } from './Coin';
