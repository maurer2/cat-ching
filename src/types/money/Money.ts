/* eslint-disable no-underscore-dangle */
/* eslint-disable @typescript-eslint/lines-between-class-members */
export const currencies = {
  GBP: 'Â£',
  USD: '$',
} as const;

export type CurrencyName = keyof typeof currencies;

export default class Money {
  private readonly _value: bigint;
  private readonly _name: CurrencyName;
  private readonly _formatter: Intl.NumberFormat;

  public static fromBigInt(
    valueInFractions: bigint,
    name: CurrencyName,
    formatter?: Intl.NumberFormat,
  ): Money {
    return new this(valueInFractions, name, formatter);
  }

  public static fromNumber(
    valueInFractions: number,
    name: CurrencyName,
    formatter?: Intl.NumberFormat,
  ): Money {
    const valueAsBigInt = BigInt(valueInFractions);

    return new this(valueAsBigInt, name, formatter);
  }

  public static fromRandom(name: CurrencyName, formatter?: Intl.NumberFormat): Money {
    const randomNumber: number = Math.ceil(Math.random() * (1000 - 1)) + 1;
    const valueAsBigInt = BigInt(randomNumber);

    return new this(valueAsBigInt, name, formatter);
  }

  private constructor(valueInFractions: bigint, name: CurrencyName, formatter?: Intl.NumberFormat) {
    this._value = valueInFractions;
    this._name = name;
    this._formatter = formatter
      ?? new Intl.NumberFormat('en-GB', {
        style: 'currency',
        currency: 'GBP',
        useGrouping: false, // disable thousand separator
      });
  }

  public add(valueToAdd: Money): Money {
    const newValue: bigint = this._value + valueToAdd._value;

    return new Money(newValue, this._name, this._formatter);
  }

  public isEqualTo(secondValue: Money): boolean {
    return this._value === secondValue._value;
  }

  public isNegative(): boolean {
    return this._value < 0;
  }

  public get valueAsSeparateParts(): Intl.NumberFormatPart[] {
    // todo Number.MAX_SAFE_INTEGER and Number.MIN_SAFE_INTEGER check
    const valueAsNumber = Number(this._value) / 100;

    return this._formatter.formatToParts(valueAsNumber);
  }

  public get valueAsFormattedString(): string {
    // todo Number.MAX_SAFE_INTEGER and Number.MIN_SAFE_INTEGER check
    const valueAsNumber = Number(this._value) / 100;

    return this._formatter.format(valueAsNumber);
  }

  public get valueFormattedAsIntegerAndFraction(): string {
    // https://stackoverflow.com/a/70086020
    const integer = Number(this._value / 100n);
    const fraction = Number(this._value % 100n);
    // const integral = Number(this.value / 100n);
    // const fraction = Number(this.value % 100n);

    return `${integer}.${`${fraction}`.padEnd(2, '0')}`;
  }
}

export type MoneyType = typeof Money;
