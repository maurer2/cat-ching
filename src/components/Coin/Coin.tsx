/* eslint-disable react/prop-types */
import React, { useState, useId } from 'react';

import Money from '../../types/Money';

import style from './Coin.module.scss';

function Button({ handleOnClick, children }) {
  return (
    <button
      className={style.button}
      onClick={handleOnClick}
      type="button"
    >
      {children}
    </button>
  );
}

// temp
const largestWidth = 28.4;
const calculateRelativeWidth = (width) => (width * 100) / largestWidth;

function Coin({ name, image, amount, size, handleAmountChange }) {
  const [showHint, setShowHint] = useState<boolean>(false);
  const currentId: string = useId();
  const targetValue = amount as Money;

  const width = calculateRelativeWidth(size.width).toFixed(2);

  function addAmount(): void {
    handleAmountChange(targetValue);
  }

  // function removeAmount(): void {
  // handleAmountChange(targetValue * -1);
  // }

  function toggleHintVisibility(): void {
    setShowHint(!showHint);
  }

  return (
    <fieldset className={style.container}>
      <label
        className={style.header}
        htmlFor={currentId}
      >
        {showHint && <span className={style.title}>{name}</span>}
        <img
          className={style.image}
          src={`/images/${image}`}
          style={{
            width: `${width}%`,
          }}
          alt=""
        />
      </label>
      <div className={style.buttonGroup}>
        <Button handleOnClick={addAmount}>Add amount</Button>

        <Button handleOnClick={toggleHintVisibility}>{showHint ? 'Hide hint' : 'Show hint'}</Button>
      </div>
    </fieldset>
  );
}

export default Coin;
