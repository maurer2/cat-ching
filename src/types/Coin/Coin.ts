import Money from '../Money';

type Coin = {
  name: string,
  image: string,
  amount: Money,
  size: {
    width: string,
    height: string,
    unit: string,
  },
};

export default Coin;
