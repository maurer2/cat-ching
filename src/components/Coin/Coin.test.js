import React from 'react';
import { shallow } from 'enzyme';
import uuidv1 from 'uuid/v1';

import Coin from './Coin';
import Money from '../../data/money';

jest.mock('uuid/v1');

describe('Coin', () => {
  const mockedHandleAmountChange = jest.fn();
  const moneyData = new Money(1, 'Dollar');

  uuidv1.mockImplementation(() => 'test-id');

  const coinData = {
    name: '1 dollar',
    image: 'image name',
    amount: moneyData,
    handleAmountChange: mockedHandleAmountChange,
    key: '1 dollar',
  }

  it('should render correctly', () => {
    const wrapper = shallow(<Coin
      {...coinData}
    />);

    expect(wrapper).toMatchSnapshot();
  });
});
