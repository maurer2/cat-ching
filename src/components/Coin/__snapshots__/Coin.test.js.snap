// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coin should render correctly 1`] = `
<fieldset
  className="container"
>
  <label
    className="header"
    htmlFor="test-id"
  >
    <img
      alt=""
      className="image"
      src="/images/image name"
    />
  </label>
  <div
    className="buttonGroup"
  >
    <Button
      handleOnClick={[Function]}
      id="test-id"
    >
      Add amount
    </Button>
    <Button
      handleOnClick={[Function]}
    >
      Remove amount
    </Button>
    <Button
      handleOnClick={[Function]}
    >
      Show hint
    </Button>
  </div>
</fieldset>
`;
